Feature/bluetooth

Wodann opened the pull request:
Extension of PR #604 that adds the ws2bth.h which is necessary for making a Windows implementation of Bluetooth sockets.

Credits to @Jarcho for all of the initial work.

Commit:
Add bthdef, bthioctl & bthsdpdef

Commit:
add bluetoothapis.rs

Commit:
Add bthledef

Commit:
Add bluetoothleapis

Commit:
Formatting fixes

Commit:
Add struct tests

Commit:
Added ws2bth

Commit:
Fixes to comply with maximum line length of 99 characters

Wodann commented:
Rebased PR to include latest commits on `0.3` branch. Can you look at this PR @retep998 ?

Wodann commented:
Rebased again to incorporate v0.3.7 release. The release of my crate is being blocked by this PR. @retep998 If you don't think it is possible to review this PR in the short-term, I'll just have to publish my forked version of the winapi crate. I've investigated other options, but there is unfortunately no other way..

retep998 commented on src/shared/bthdef.rs:
```diff
@@ -0,0 +1,1127 @@
+// Copyright © 2018 winapi-rs developers
```
The copyright lines were removed from `winapi` recently and should be removed from this PR.

retep998 commented on src/shared/bthdef.rs:
```diff
@@ -0,0 +1,1127 @@
+pub const UDI_C_PLANE_PROTOCOL_UUID16: USHORT = 0x001D;
+pub const L2CAP_PROTOCOL_UUID16: USHORT = 0x0100;
+DEFINE_BLUETOOTH_UUID128!{SDP_PROTOCOL_UUID,
+    SDP_PROTOCOL_UUID16}
```
All these `DEFINE_BLUETOOTH_UUID128!` invocations should be single line invocations.

retep998 commented on src/shared/bthdef.rs:
```diff
@@ -0,0 +1,1127 @@
+    IoCaps_DisplayYesNo    = 0x01,
+    IoCaps_KeyboardOnly    = 0x02,
+    IoCaps_NoInputNoOutput = 0x03,
+    IoCaps_Undefined       = 0xff,
```
No aligned indentation.

retep998 commented on src/um/bluetoothapis.rs:
```diff
@@ -0,0 +1,415 @@
+use shared::windef::HWND;
+use um::minwinbase::SYSTEMTIME;
+use um::winnt::{
+    CHAR, HANDLE, LONG, LONGLONG, LPCWSTR, LPWSTR, PVOID, PWSTR, SHORT, ULONGLONG, WCHAR
```
Trailing comma is missing.

retep998 commented on src/um/bluetoothapis.rs:
```diff
@@ -0,0 +1,415 @@
+}}
+pub type PBLUETOOTH_DEVICE_INFO = *mut BLUETOOTH_DEVICE_INFO;
+ENUM!{enum BLUETOOTH_AUTHENTICATION_METHOD {
+    BLUETOOTH_AUTHENTICATION_METHOD_LEGACY     = 0x1,
```
Extra spaces that shouldn't be there.

retep998 commented on src/um/bluetoothapis.rs:
```diff
@@ -0,0 +1,415 @@
+    BLUETOOTH_IO_CAPABILITY_DISPLAYYESNO    = 0x01,
+    BLUETOOTH_IO_CAPABILITY_KEYBOARDONLY    = 0x02,
+    BLUETOOTH_IO_CAPABILITY_NOINPUTNOOUTPUT = 0x03,
+    BLUETOOTH_IO_CAPABILITY_UNDEFINED       = 0xff,
```
No aligned indentation.

retep998 commented on src/um/bluetoothapis.rs:
```diff
@@ -0,0 +1,415 @@
+    BLUETOOTH_MITM_ProtectionRequiredBonding    = 0x3,
+    BLUETOOTH_MITM_ProtectionNotRequiredGeneralBonding    = 0x4,
+    BLUETOOTH_MITM_ProtectionRequiredGeneralBonding    = 0x5,
+    BLUETOOTH_MITM_ProtectionNotDefined            = 0xff,
```
What even.

retep998 commented on src/um/bluetoothleapis.rs:
```diff
@@ -0,0 +1,105 @@
+    BLUETOOTH_GATT_EVENT_HANDLE, BTH_LE_GATT_EVENT_TYPE, BTH_LE_GATT_RELIABLE_WRITE_CONTEXT,
+    PBTH_LE_GATT_CHARACTERISTIC, PBTH_LE_GATT_CHARACTERISTIC_VALUE, PBTH_LE_GATT_DESCRIPTOR,
+    PBTH_LE_GATT_DESCRIPTOR_VALUE, PBTH_LE_GATT_RELIABLE_WRITE_CONTEXT, PBTH_LE_GATT_SERVICE,
+    PFNBLUETOOTH_GATT_EVENT_CALLBACK
```
Missing trailing comma.

retep998 commented on src/um/bthledef.rs:
```diff
@@ -0,0 +1,282 @@
+pub const BTH_LE_ATT_MAX_VALUE_SIZE: USHORT = 512;
+pub const BTH_LE_ATT_CID: USHORT = 0x0004;
+pub const BTHLEENUM_ATT_MTU_MIN: USHORT = 23;
+pub const BTHLEENUM_ATT_MTU_MAX: USHORT = u16::MAX;
```
The value of this should be `MAX_USHORT` but since that is apparently never defined anywhere, this should just be commented out.

retep998 commented on src/um/ws2bth.rs:
```diff
@@ -0,0 +1,212 @@
+// All files in the project carrying such notice may not be copied, modified, or distributed
+// except according to those terms.
+use shared::bthdef::{
+    BTH_ADDR, MAX_L2CAP_INFO_DATA_LENGTH, MAX_L2CAP_PING_DATA_LENGTH, MAX_UUIDS_IN_QUERY
```
Trailing comma.

retep998 commented on src/um/ws2bth.rs:
```diff
@@ -0,0 +1,212 @@
+    addressFamily: USHORT,  // Always AF_BTH
+    btAddr: BTH_ADDR,       // Bluetooth device address
+    serviceClassId: GUID,   // [OPTIONAL] system will query SDP for port
+    port: ULONG,            // RFCOMM channel or L2CAP PSM
```
Remove these comments.

retep998 commented on src/um/ws2bth.rs:
```diff
@@ -0,0 +1,212 @@
+pub type PSOCKADDR_BTH = *mut SOCKADDR_BTH;
+DEFINE_GUID!{SVCID_BTH_PROVIDER,
+    0x6aa63e0, 0x7d60, 0x41ff, 0xaf, 0xb2, 0x3e, 0xe6, 0xd2, 0xd9, 0x39, 0x2d}
+pub const BT_ADDR_STRING_SIZE: DWORD = 12; // max size from WSAAddressToString
```
This is actually `BTH_ADDR_STRING_SIZE`. Also remove the comment.

retep998 commented on src/um/ws2bth.rs:
```diff
@@ -0,0 +1,212 @@
+pub const SO_BTH_MTU_MIN: ULONG = 0x8000000a;
+// 3-DH5 => payload of 1021 => L2cap payload of 1017 => RFComm payload of 1011
+pub const RFCOMM_MAX_MTU: ULONG = 0x000003F3;  // L2CAP MTU (1017) - RFCOMM header size (6)
+pub const RFCOMM_MIN_MTU: ULONG = 0x00000017;  // RFCOMM spec sec 5.3 table 5.1
```
Just remove all these extraneous comments please.

Commit:
Removed copyright notices from bluetooth filees

Commit:
Compacted two-line statement to a single line

Commit:
Added trailing commas to use blocks

Commit:
Removed aligned indentation

Commit:
Removed documentation comments

Commit:
Removed undefined assignment

Wodann commented:
Thank you for your review @retep998. I fixed all of the issues you pointed out, and did the same in places where I found similar issues. I hope the updated PR is up to your standards.

I also rebased the PR on top of your latest commit in the `0.3` branch again to prevent potential merge conflicts.

Commit:
Removed unused use statement
